### ISSUE TITLE
Feature complete through odbc version 3.8 

### ISSUE THREAD

**Issue by auroraeosrose** [SammyK/php-src-odbc#1]
Microsoft is actually the best place for good documentation on odbc and standards

https://msdn.microsoft.com/en-us/library/ms714177.aspx

We need to 

1. catalog the odbc functionality currently available in the php odbc extension
2. map to existing functionality in the odbc standard and find gaps
3. check existing apis for silly api design
4. work on OO version of the api design
5. add some of the more complex functionality available through other extensions (traversable statements, lazy pull, etc)

The point of course is NOT just to "rewrite odbc to use resources" although that IS the main goal
it's also to 
add an object oriented api
add features available in newer odbc versions
cleanup and make the api nicer
add features that other db extensions enjoy
cleanup the code - dropping support for things like birdstep/solid/other crazy ifdefs until we're sure they're necessary (I think birdstep, for example, should be in it's own extension)

Also verify all the different odbc versions supported - right now the ifdefs in the headers for odbc seem quite silly but it looks like welting has been on a tear cleaning up odbc as best as can be done without a rewrite

**Comment by auroraeosrose** [comment]
So these are the current odbc functions support in PHP 7.1
that I'd like to allow aliases for in odbci when compiled appropriately (with-odbc-alias or something)

const zend_function_entry odbc_functions[] = {
	PHP_FE(odbc_autocommit, arginfo_odbc_autocommit)
	PHP_FE(odbc_binmode, arginfo_odbc_binmode)
	PHP_FE(odbc_close, arginfo_odbc_close)
	PHP_FE(odbc_close_all, arginfo_odbc_close_all)
	PHP_FE(odbc_columns, arginfo_odbc_columns)
	PHP_FE(odbc_commit, arginfo_odbc_commit)
	PHP_FE(odbc_connect, arginfo_odbc_connect)
	PHP_FE(odbc_cursor, arginfo_odbc_cursor)
#ifdef HAVE_SQLDATASOURCES
	PHP_FE(odbc_data_source, arginfo_odbc_data_source)
#endif
	PHP_FE(odbc_execute, arginfo_odbc_execute)
	PHP_FE(odbc_error, arginfo_odbc_error)
	PHP_FE(odbc_errormsg, arginfo_odbc_errormsg)
	PHP_FE(odbc_exec, arginfo_odbc_exec)
#ifdef PHP_ODBC_HAVE_FETCH_HASH
	PHP_FE(odbc_fetch_array, arginfo_odbc_fetch_array)
	PHP_FE(odbc_fetch_object, arginfo_odbc_fetch_object)
#endif
	PHP_FE(odbc_fetch_row, arginfo_odbc_fetch_row)
	PHP_FE(odbc_fetch_into, arginfo_odbc_fetch_into)
	PHP_FE(odbc_field_len, arginfo_odbc_field_len)
	PHP_FE(odbc_field_scale, arginfo_odbc_field_scale)
	PHP_FE(odbc_field_name, arginfo_odbc_field_name)
	PHP_FE(odbc_field_type, arginfo_odbc_field_type)
	PHP_FE(odbc_field_num, arginfo_odbc_field_num)
	PHP_FE(odbc_free_result, arginfo_odbc_free_result)
	PHP_FE(odbc_gettypeinfo, arginfo_odbc_gettypeinfo)
	PHP_FE(odbc_longreadlen, arginfo_odbc_longreadlen)
#if !defined(HAVE_SOLID) && !defined(HAVE_SOLID_30)
	PHP_FE(odbc_next_result, arginfo_odbc_next_result)
#endif
	PHP_FE(odbc_num_fields, arginfo_odbc_num_fields)
	PHP_FE(odbc_num_rows, arginfo_odbc_num_rows)
	PHP_FE(odbc_pconnect, arginfo_odbc_pconnect)
	PHP_FE(odbc_prepare, arginfo_odbc_prepare)
	PHP_FE(odbc_result, arginfo_odbc_result)
	PHP_FE(odbc_result_all, arginfo_odbc_result_all)
	PHP_FE(odbc_rollback, arginfo_odbc_rollback)
	PHP_FE(odbc_setoption, arginfo_odbc_setoption)
	PHP_FE(odbc_specialcolumns, arginfo_odbc_specialcolumns)
	PHP_FE(odbc_statistics, arginfo_odbc_statistics)
	PHP_FE(odbc_tables, arginfo_odbc_tables)
	PHP_FE(odbc_primarykeys, arginfo_odbc_primarykeys)
#if !defined(HAVE_DBMAKER) && !defined(HAVE_SOLID) && !defined(HAVE_SOLID_30) &&!defined(HAVE_SOLID_35) && !defined(HAVE_BIRDSTEP)    /* not supported now */
	PHP_FE(odbc_columnprivileges, arginfo_odbc_columnprivileges)
	PHP_FE(odbc_tableprivileges, arginfo_odbc_tableprivileges)
#endif
#if !defined(HAVE_SOLID) && !defined(HAVE_SOLID_30) && !defined(HAVE_SOLID_35) /* not supported */
	PHP_FE(odbc_foreignkeys, arginfo_odbc_foreignkeys)
	PHP_FE(odbc_procedures, arginfo_odbc_procedures)
#if !defined(HAVE_BIRDSTEP)
	PHP_FE(odbc_procedurecolumns, arginfo_odbc_procedurecolumns)
#endif
#endif
	PHP_FALIAS(odbc_do, odbc_exec, arginfo_odbc_exec)
	PHP_FALIAS(odbc_field_precision, odbc_field_len, arginfo_odbc_field_len)
	PHP_FE_END
};

**Comment by auroraeosrose** [comment]
One thing we might want to think about is the state of connection pooling/persistent connections
BECAUSE odbc has connection pooling we really should make sure we're using that as much as possible - pdo_odbc already does the connection pooling right - the nasty way odbc currently does "persistent" handles needs to die 

(note, this might have small ramifications for bc - but I believe it would actually make older apps using persistent functionality to behave BETTER since there would be better control over the pooling since it's at the odbc driver manager level instead of at nasty php persist the resource pointer level)

https://msdn.microsoft.com/en-us/library/hh405036.aspx has a bunch of heavy reading about it (it's rather clever actually)

https://msdn.microsoft.com/en-us/library/ms709285(v=vs.85).aspx has more on not only the old method of connection pooling but the new stuff as well

**Comment by auroraeosrose** [comment]
we're going to need a heck of a lot more tests too
there are exactly 13 tests for the odbc extension HALF of which are bugs :(

**Comment by auroraeosrose** [comment]
Yeah, after some research definitely adding "kill persistent connections and add proper use of odbc connection pooling instead" to the list

**Comment by SammyK** [comment]
Started to work on the API a little bit today: https://github.com/SammyK/php-src-odbci/wiki

**Comment by auroraeosrose** [comment]
I'd argue we don't want an odbc_close_all - that was in really in use because of the terribad implementation of "persistent" connections - instead we want to allow usage of odbc connection pools instead

yes I'd stick the error message and error code into exception, let's NOT play the pdo game with that :) for the "functional" api we can have functions to pull that out

I think that the odbci result object (like pdo) should be more or less our "default" object fetch, but like pdo we should allow injecting a class name as into the object fetch

So a set of hte functionality we see in odbc are called "catalog functions"
it might be wise to move these into their own class

https://msdn.microsoft.com/en-us/library/ms711722(v=vs.85).aspx

these are the ones that you can use to introspect your table layouts, etc

there are also a set of interesting cursor and bulk insert related items in odbc that just aren't supported

**Comment by SammyK** [comment]
Sounds good!

> for the "functional" api we can have functions to pull that out

Can we just have the functions throw `odbciException` instead of returning false and trying to get the error from another function?

**Comment by SammyK** [comment]
Even more updates :)

https://github.com/SammyK/php-src-odbci/wiki
